### DISCUSSION TITLE
Candidate ID

### DISCUSSION THREAD

**Discussion by drpeterallan** [anyoptimization/pymoo#211] [Q&A] [answered]
I would like to know does pymoo have a unique ID for each candidate within a population? For my problem I have a pool of candidates and I would like to uniquely identify these and understand at which generation a candidate was produced. For instance if a candidate solution was produced at generation 10 say and this survives to the next gen, how could I differentiate between these 2 candidates? I am saving every candidate from each generation to a csv. I understand I could do this manually by checking candidate parameters but I just wondered if pymoo assigns something like a UID and if this is preserved for generation to generation. For reference I am using version 0.5.0 and NSGA-III

Thanks in advance

**Comment by blankjul** [comment, marked as answer]
you can use the Python `id` method. So for each individual `i` do `id(algorithm.pop[i])` which gives you an id. You can append this to your CSV data and then reconstruct the run. 
Does that work for you?

**Reply by drpeterallan** [reply]
That may work thank you, I'll take a look. 

I have a follow up question please, is it possible to save the state of the algorithm at each generation, i.e. set save_history=False and just save this at each iteration?

**Comment by blankjul** [comment]
Yes! Check out `Callback`.

https://pymoo.org/interface/callback.html

**Reply by drpeterallan** [reply]
OK, I've made some changes as you suggested regarding the individual id and I'm seeing somewhat unexpected behaviour. As mentioned I'm using NSGA-III, the setup of my algorithm is as follows:

 `algorithm = NSGA3(pop_size=165, ref_dirs=ref_dirs,     sampling=get_sampling("real_random"), cross_over=get_crossover("real_sbx", 0.5), mutation=get_mutation("real_pm", prob=0.5))` 

and `ref_dirs = get_reference_directions("das-dennis", 3, n_partitions=8)`. This runs for the specified number of generations and for each generation I get 165 individuals which are completely unique. I assumed post nondominated sorting and candidate generation that some of the individuals would survive to the next generation, this doesn't appear to be the case? Looking further into the documentation I decided to set `n_offsprings=10` assuming this would generate 10 new individuals and the remaining 155 individuals would survive to the next generation. 10 new individuals are generated each generation but when I look at the `res.history`, firstly each gen has a population of 165 and when I compare gen 1 vs. 2 it shows that there are 18 nonunique individuals, i.e. they have survived. I'm not clear on what is occurring here. 

I suppose my question is how do I set something equivalent to a memory in pymoo, i.e. I want to preserve the 40 best candidates in a generation say and then create new candidates based off of those individuals?

I really appreciate your help on this

**Comment by blankjul** [comment]
If you want to keep the 40 best candidates then you have to set `pop_size` equal to 40. However, because NSGA-III is reference-direction based you should find a partition number first which creates close to 40 individuals. If this is the case, then there is no need to provide a population size.

So I would use `get_reference_directions("das-dennis", 3, n_partitions=8)` which creates 45 reference directions. If you do not provide any `pop_size` it will be set to 45 as well. If you then set `n_offsprings=10` then in each iteration 10 offsprings are created through mating. The merged population will have 55 individuals, and the survival will again reduce them to 10.

**Reply by drpeterallan** [reply]
OK, so If I set `n_offsprings` to none and have a population of 165 with `ref_dirs = get_reference_directions("das-dennis", 3, n_partitions=8)` then each generation I get 165 new candidates? So no individuals survive to the next generation?

**Comment by blankjul** [comment]
if `n_offsprings` is not set it will be equal to the `pop_size`. This means your `pop` has 165, then you create 165 offsprings. This results in a merged population of 165 + 165 = 330. From the merged population, 165 solutions will be selected to survive. It is an elitist environment survival (I am not sure what you trying to say with your last question).

**Reply by drpeterallan** [reply]
OK that clears things up for me thank you
